### PR TITLE
Delay loading the dev cert 

### PR DESCRIPTION
**Pull request by Tratcher**
#2422 Loading the dev cert is now delayed until after all possible user code that would set a cert.

### CODE REVIEW

**Comment by halter73** [comment]
Why not just delay until the first request? I agree with what @davidfowl said in the issue. Why not just use the SNI callback to load the default cert? I like there being less special cases. It's not like you have to go to disk for every connection. You can cache the result.

File: src/Kestrel.Core/KestrelServerOptions.cs [hunk 1]
```diff
@@ -105,10 +115,50 @@ public void ConfigureHttpsDefaults(Action<HttpsConnectionAdapterOptions> configu
 
         internal void ApplyHttpsDefaults(HttpsConnectionAdapterOptions httpsOptions)
         {
-            httpsOptions.ServerCertificate = DefaultCertificate;
             HttpsDefaults(httpsOptions);
         }
 
+        internal void ApplyDefaultCert(HttpsConnectionAdapterOptions httpsOptions)
```
Comment (halter73) [inline src/Kestrel.Core/KestrelServerOptions.cs]: Nit: Rename to TryApplyDefaultCert of EnsureDefaultCert.

Comment (Tratcher) [reply src/Kestrel.Core/KestrelServerOptions.cs]: I thought about that but you'd normally use Try when you were going to return the status and act differently. Here it just no-ops.

**Comment by Tratcher** [comment]
Rebased onto the SNI changes and added a matching check to not load the dev cert if the SNI callback is given.

**Comment by halter73** [comment]
Still LGTM after rebase.

File: src/Kestrel.Core/KestrelServerOptions.cs (hunk 1 above)
Comment (halter73) [reply src/Kestrel.Core/KestrelServerOptions.cs]: EnsureServerCert it is then?

**Pull request merged**